Added request size limit value to exception message in case of request body large exception for Kestrel

ShreyasJejurkar opened the pull request:
Fixes https://github.com/dotnet/aspnetcore/issues/4773

ShreyasJejurkar committed:
Added request size limit value to exception message in case of request body large exception for kestral.

Fixes https://github.com/dotnet/aspnetcore/issues/4773

davidfowl commented on src/Servers/Kestrel/Core/src/CoreStrings.resx:
```diff
@@ -307,7 +307,7 @@
   </data>
   <data name="BadRequest_RequestBodyTooLarge" xml:space="preserve">
-    <value>Request body too large.</value>
+    <value>Request body too large. Default value is 30,000,000 bytes, which is approximately 28.6MB. </value>
```
We shouldn't hard code this in the exception. We should pass the actual limit here, not a hardcoded limit.

MCCshreyas replied:
Ohh make sense. I will get that fixed!

MCCshreyas replied:
But the problem is how could I get to `_maxRequestBodySize` (or its public property) which is present here https://github.com/dotnet/aspnetcore/blob/4cc933927fb668a67e1ce6f2e368edb45af164b4/src/Servers/Kestrel/Core/src/KestrelServerLimits.cs#L31

in static class here 
https://github.com/dotnet/aspnetcore/blob/4cc933927fb668a67e1ce6f2e368edb45af164b4/src/Servers/Kestrel/Core/src/KestrelBadHttpRequestException.cs#L68

I tried to find

davidfowl replied:
An example https://github.com/dotnet/aspnetcore/blob/7a20550cc61bd7661836d69f071919e6850e7867/src/Servers/Kestrel/Core/src/CoreStrings.resx#L211. Plus you need to re-run a command on the command line (remind me @javiercn) to generate a new method (it will be a method because it has arguments)

MCCshreyas replied:
Oh, but we need to pass `requestBufferSize` with `string.Format` correct? Or not?

MCCshreyas replied:
They have access to `KestralOptions` here from where they are passing those two values!
https://github.com/dotnet/aspnetcore/blob/ea72e31a4dc3b332887533a0d7b0dd5917f3782e/src/Servers/Kestrel/Core/src/Internal/KestrelServerImpl.cs#L367-L381

halter73 replied:
@MCCshreyas Everywhere `KestrelBadHttpRequestException.Throw(RequestRejectionReason.RequestBodyTooLarge)` is called, the `MaxRequestBodySize` is accessible. Otherwise the code wouldn't know whether or not to throw.

There's both `Throw(RequestRejectionReason)` and `Throw(RequestRejectionReason, string)`. You can move `RequestRejectionReason.RequestBodyTooLarge` to [the overload that takes `string detail`](https://github.com/dotnet/aspnetcore/blob/c925f99cddac0df90ed0bc4a07ecda6b054a0b02/src/Servers/Kestrel/Core/src/KestrelBadHttpRequestException.cs#L100) and pass the `MaxRequestBodySize` as a string.

Here are the call sites for `KestrelBadHttpRequestException.Throw(RequestRejectionReason.RequestBodyTooLarge)` if you still want to take a crack at fixing this:

- https://github.com/dotnet/aspnetcore/blob/c925f99cddac0df90ed0bc4a07ecda6b054a0b02/src/Servers/Kestrel/Core/src/Internal/Http/MessageBody.cs#L178
- https://github.com/dotnet/aspnetcore/blob/c925f99cddac0df90ed0bc4a07ecda6b054a0b02/src/Servers/Kestrel/Core/src/Internal/Http/Http1ContentLengthMessageBody.cs#L240
- https://github.com/dotnet/aspnetcore/blob/c925f99cddac0df90ed0bc4a07ecda6b054a0b02/src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs#L29
- https://github.com/dotnet/aspnetcore/blob/c925f99cddac0df90ed0bc4a07ecda6b054a0b02/src/Servers/Kestrel/Core/src/Internal/Http3/Http3MessageBody.cs#L29

halter73 commented:
@MCCshreyas do you think you'll have time to come back to this? If not, no worries.

MCCshreyas commented:
Hi @halter73, Sorry for the delay! 
I will get this done and checked this weekend! 

Thanks!

MCCshreyas commented:
@davidfowl Can you please tell me the command via which the resource method gets autogenerated with parameter? I have updated the resource string with the parameter, now I need to generate its corresponding method! 

Thanks!

davidfowl commented:
> @davidfowl Can you please tell me the command via which the resource method gets autogenerated with parameter? I have updated the resource string with the parameter, now I need to generate its corresponding method!

@dougbu can you help out here? Is there a doc on how you generate resx? I recall it was build.cmd resx but I don't think there's a build in this solution folder anymore

dougbu commented:
> Is there a doc on how you generate resx?

See https://github.com/dotnet/aspnetcore/blob/main/docs/BuildFromSource.md#resx-files

Note the generated files aren't usually checked in and the resources are compiled during builds.

davidfowl commented:
@MCCshreyas are you going to finish this?

davidfowl commented:
@MCCshreyas are you going to finish this change?

MCCshreyas commented:
Yeah, this weekend.

MCCshreyas commented:
@dougbu Not sure why I am getting this error. Am running this command from the same folder where the `csproj` is

![image](https://user-images.githubusercontent.com/17148381/114262285-c4f30000-99fc-11eb-8f6e-b5fbd0de8045.png)

davidfowl commented:
Your command looks wrong. Do this:

```
dotnet msbuild /t:Resgen
```

You have typo: `t:/Resgen` instead of `/t:Resgen`, you also don't need to specify the project file, that folder only has one.

MCCshreyas commented:
Thanks, @davidfowl, It worked. Seems like a typo in docs! Lol. I will get that fixed that as well!

MCCshreyas commented:
Hi, @davidfowl, How can I instruct `resx` generator to generate a method for that and not property. I have added the below message with parameters, but the default behavior is generating `property` and not the method. 
![image](https://user-images.githubusercontent.com/17148381/114262784-81e65c00-99ff-11eb-8085-02be7dd4a3f8.png)

![image](https://user-images.githubusercontent.com/17148381/114262797-9cb8d080-99ff-11eb-9242-0de376ee8ad4.png)

ShreyasJejurkar committed:
1. Added paramters to resource string and generated corresponding C# method.
2. Changed the callsite, so that we can pass maxRequestBodySize to exception method.

ShreyasJejurkar committed:
Merge branch 'main' into kestral-message-fix

davidfowl commented:
```
##[error]src/Servers/Kestrel/Core/src/Internal/Http/MessageBody.cs(179,98): error CA1305: (NETCORE_ENGINEERING_TELEMETRY=Build) The behavior of 'long.ToString()' could vary based on the current user's locale settings. Replace this call in 'MessageBody.AddAndCheckObservedBytes(long)' with a call to 'long.ToString(IFormatProvider)'.
```

MCCshreyas commented:
@davidfowl, I was looking for example, on how people have passed `string` argument to this one, but they were having an argument in `string` type already, but in my case, `MaxBodyRequestSize` is in `long?`, so anyhow I need to convert that before passing! 

So to pass the instance of `IFormatProvider` should I take a constructor dependency for this or the method dependency? Any example for me to refer to in the existing codebase! 

Thanks!

davidfowl commented:
InvariantCulture is what you want

ShreyasJejurkar committed:
Added CultureInfo.InvariantCulture.

1. Passed to ToString() while converting from `long?` to `string`.

ShreyasJejurkar committed:
Merge branch 'kestral-message-fix' of https://github.com/MCCshreyas/aspnetcore into kestral-message-fix

halter73 committed:
Merge branch 'main' into kestral-message-fix

halter73 committed:
Fix tests and update exception message to not reference "default"

halter73 reviewed (approved):
Thanks!

halter73 commented:
@MCCshreyas I also saw your [reddit post](https://www.reddit.com/r/dotnet/comments/mup6bm/minimal_hosting_making_its_way_to_aspnet_core/) about minimal hosting. Thanks for putting the word out!

MCCshreyas commented:
No worries, @halter73. Just putting some initial awareness about new features! 😅😅
Atleast I can do that! 😇😇
